server/src/main/java/com/angelomelonas/grpcwebchat/service/ChatSessionService.java

```diff
@@ -0,0 +1,108 @@
+
+public class ChatSessionService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ChatSessionService.class);
+
+    ConcurrentHashMap<UUID, ChatSession> chatSessionHashMap;
+
+    public ChatSessionService() {
+        this.chatSessionHashMap = new ConcurrentHashMap<>();
+    }
+
+    public void create(UUID uuid, StreamObserver<AuthenticationResponse> responseObserver) {
+        ChatSession chatSession = new ChatSession();
+
+        // Add the ChatSession.
+        this.chatSessionHashMap.put(uuid, chatSession);
+
+        final AuthenticationResponse authenticationResponse = AuthenticationResponse.newBuilder()
+                .setUuid(String.valueOf(uuid))
+                .build();
+
+        // Respond with the newly generated UUID.
+        responseObserver.onNext(authenticationResponse);
+        responseObserver.onCompleted();
+
+        LOGGER.info("ChatSession with UUID {} created.", uuid);
+    }
+
+    public void subscribe(UUID uuid, String username, StreamObserver<Message> responseObserver) {
+        ChatSession chatSession = chatSessionHashMap.get(uuid);
+
+        if (chatSession == null) {
+            IllegalArgumentException illegalArgumentException = new IllegalArgumentException("Cannot subscribe. Client not authenticated.");
+            responseObserver.onError(illegalArgumentException);
+            throw illegalArgumentException;
+        }
+
+        chatSession.subscribe(username, responseObserver);
+
+        final Message clientSubscribed = Message.newBuilder()
+                .setUsername("Server")
+                .setMessage("Client subscribed successfully.")
+                .setTimestamp(Instant.now().toEpochMilli())
+                .build();
+
+        responseObserver.onNext(clientSubscribed);
+
+        LOGGER.info("ChatSession with UUID {} and username {} subscribed.", uuid, username);
+    }
+
+    public void unsubscribe(UUID uuid, StreamObserver<Message> responseObserver) {
+        ChatSession chatSession = this.chatSessionHashMap.remove(uuid);
+
+        if (chatSession == null) {
+            IllegalArgumentException illegalArgumentException = new IllegalArgumentException("Cannot unsubscribe. Client not subscribed.");
+            responseObserver.onError(illegalArgumentException);
+            throw illegalArgumentException;
+        }
+
+        chatSession.unsubscribe();
+
+        final Message clientUnsubscribed = Message.newBuilder()
+                .setUsername("Server")
+                .setMessage("You have successfully been unsubscribed")
+                .setTimestamp(Instant.now().toEpochMilli())
+                .build();
+
+        responseObserver.onNext(clientUnsubscribed);
+        responseObserver.onCompleted();
+
+        LOGGER.info("ChatSession with UUID {} unsubscribed.", uuid);
+    }
+
+    public synchronized void sendMessage(UUID uuid, String username, String message, StreamObserver<Message> responseObserver) {
```

Comment:
just going though this class ... it feels that the responseObserver should not have to be passed in each time. Should that not be stored within each ClientSession

Reply:
maybe the method should just be:
sendMessage(UUID sessionID, String username, String message)

even better:
sendMessage(UUID sessionID, String message)

Reply:
Just to be sure, you are aware that those observers being passed in are the blocking observers that are closed immediately? 

With that said, I wanted to keep the gRPC logic in the Service class. However, do you think it's better to then respond with the success message from within the ChatService (API) class?